pkg/ruler/rulestore/bucket_client.go

```diff
@@ -0,0 +1,312 @@
+package rulestore
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	"github.com/cortexproject/cortex/pkg/tenant"
+)
+
+const (
+	delim      = "/"
+	rulePrefix = "rules"
+)
+
+type BucketRuleStore struct {
+	bucket          objstore.Bucket
+	loadConcurrency int
+
+	logger log.Logger
+}
+
+func NewBucketRuleStore(bucket objstore.Bucket, loadConcurrency int, logger log.Logger) *BucketRuleStore {
+	return &BucketRuleStore{
+		bucket:          bucket,
+		loadConcurrency: loadConcurrency,
+		logger:          logger,
+	}
+}
+
+func newRuleStoreUserBucketClient(userID string, bkt objstore.Bucket) objstore.Bucket {
+	return bucket.NewUserBucketClient(userID, bucket.NewPrefixedBucketClient(bkt, rulePrefix))
+}
+
+func (b *BucketRuleStore) listNamespacesForUser(ctx context.Context, user string) ([]string, error) {
+	userBucket := newRuleStoreUserBucketClient(user, b.bucket)
+
+	var namespaces []string
+	err := userBucket.Iter(ctx, "", func(namespace string) error {
+		namespace = strings.TrimSuffix(namespace, delim)
+		decodedNamespace, err := base64.URLEncoding.DecodeString(namespace)
+		if err != nil {
+			return err
```

Comment:
```suggestion
			return errors.Wrap(err, "failed to decode namespace")
```

Reply:
Any opposition to using the standard library error wrapping instead?

```go
return fmt.Errorf("failed to decode namespace '%s', %w", namespace, err)
```

Reply:
No, although at some point it would be nice to unify this, and we use `errors` package quite heavily already.

Reply:
> No, although at some point it would be nice to unify this, and we use errors package quite heavily already.

I agree, I created a thread to center the discussion on this topic: https://github.com/cortexproject/cortex/discussions/3834